java/ql/src/Telemetry/AutomodelEndpointCharacteristics.qll

```diff
@@ -0,0 +1,325 @@
+      ExternalFlow::sinkModel(package, type, subtypes, name, [signature, ""], ext, input, label, _)
+    )
+  }
+
+  predicate isNeutral(Endpoint e) {
+    exists(string package, string type, string name, string signature |
+      sinkSpec(e, package, type, _, name, signature, _, _) and
+      ExternalFlow::neutralModel(package, type, name, [signature, ""], _)
+    )
+  }
+
+  additional predicate sinkSpec(
+    Endpoint e, string package, string type, boolean subtypes, string name, string signature,
+    string ext, string input
+  ) {
+    package = e.getEnclosingCallable().getDeclaringType().getPackage().toString() and
+    type = e.getEnclosingCallable().getDeclaringType().getName() and
+    subtypes = false and
+    name = e.getEnclosingCallable().getName() and
+    signature = ExternalFlow::paramsString(e.getEnclosingCallable()) and
+    ext = "" and
+    exists(int paramIdx | e.isParameterOf(_, paramIdx) | input = "Argument[" + paramIdx + "]")
+  }
+
+  predicate hasMetadata(Endpoint e, string metadata) {
+    exists(
+      string package, string type, boolean subtypes, string name, string signature, string ext,
+      int input, string provenance, boolean isPublic, boolean isFinal, boolean isStatic,
+      string calleeJavaDoc
+    |
+      hasMetadata(e, package, type, name, signature, input, isFinal, isStatic, isPublic,
+        calleeJavaDoc) and
+      (if isFinal = true or isStatic = true then subtypes = false else subtypes = true) and
+      ext = "" and // see https://github.slack.com/archives/CP9127VUK/p1673979477496069
+      provenance = "ai-generated" and
+      metadata =
+        "{" //
+          + "'Package': '" + package //
+          + "', 'Type': '" + type //
+          + "', 'Subtypes': " + subtypes //
+          + ", 'Name': '" + name //
+          + ", 'ParamName': '" + e.toString() //
+          + "', 'Signature': '" + signature //
+          + "', 'Ext': '" + ext //
+          + "', 'Argument index': " + input //
+          + ", 'Provenance': '" + provenance //
+          + "', 'Is public': " + isPublic //
+          + "', 'Callee JavaDoc': '" + calleeJavaDoc.replaceAll("'", "\"") //
+          + "'}" // TODO: Why are the curly braces added twice?
+    )
+  }
+}
+
+module CharacteristicsImpl = SharedCharacteristics::SharedCharacteristics<CandidatesImpl>;
+
+class EndpointCharacteristic = CharacteristicsImpl::EndpointCharacteristic;
+
+class Endpoint = CandidatesImpl::Endpoint;
+
+/*
+ * Predicates that are used to surface prompt examples and candidates for classification with an ML model.
+ */
+
+/**
+ * Holds if `n` has the given metadata.
+ *
+ * This is a helper function to extract and export needed information about each endpoint.
+ */
+predicate hasMetadata(
+  Endpoint n, string package, string type, string name, string signature, int input,
+  boolean isFinal, boolean isStatic, boolean isPublic, string calleeJavaDoc
+) {
+  exists(Callable callee |
```

Comment:
Minor: it is a bit confusing to call this the callee when we aren't actually looking at any calls in this mode. Perhaps `function`?

Reply:
`callable` is good!